Adapt for SLES4SAP

varkoly opened the pull request:
- Adapt partitioning for SLES4SAP
- Remove first_user module

Commit:
- Adapt partitioning for SLES4SAP
- Remove first_user module

jreidinger commented:
@varkoly thanks for adaptation looks good. Only two notes:

1. please increase version. Without it no submit request will be created

2. please add bnc or FATE ID to changes. There is a bot that checks it in submit requests.

Thanks

varkoly committed:
Add fate to changes.
Introduce <lvm_vg_strategy> tag to partitioning proposal.
Replace user_first by sap-start module.

lslezak commented:
# Suggested Fixes

## The Name Space Issue

To avoid adding `xmlns="" xmlns:n="http://www.suse.com/1.0/yast2ns"` attributes in the output add 
```
  xmlns="http://www.suse.com/1.0/yast2ns"
  exclude-result-prefixes="n"
```

attributes to the initial `<xsl:stylesheet>` tag.

## Removing CDATA

I noticed that the output XML contains this change:

```diff
                   <!-- the cdata trick produces an empty string in the data
                        instead of omitting the key entirely -->
-                  <optional_default_patterns><![CDATA[]]></optional_default_patterns>
+                  <optional_default_patterns></optional_default_patterns>
```
Normally these two are equivalent, however the YaST XML parser handles that a bit differently and needs that as a workaround as written in the comment above.

After some googling I found this workaround for it:
```xml
  <!-- keep empty <![CDATA[]]>, see https://stackoverflow.com/a/1364900 -->
  <xsl:template match="n:optional_default_patterns">
      <xsl:element name="optional_default_patterns">
          <xsl:text disable-output-escaping="yes">&lt;![CDATA[</xsl:text>
          <xsl:value-of select="text()" disable-output-escaping="yes"/>
          <xsl:text disable-output-escaping="yes">]]&gt;</xsl:text>
      </xsl:element>
  </xsl:template>
```

## Preserving Space in Added Tags

I also noticed that the added step in the workflow removed all whitespace:

```diff
-            <module>
-                <name>user_first</name>
-            </module>
+            <module><label>Installation Mode</label><name>sap-start</name><enable_back>yes</enable_back><enable_next>yes</enable_next></module>
```
After using `xml:space="preserve"` option in 

```xml
  <xsl:template xml:space="preserve" match="n:module[n:name='user_first']"><module>
                <label>Installation Mode</label>
                <name>sap-start</name>
                <enable_back>yes</enable_back>
                <enable_next>yes</enable_next>
            </module></xsl:template>
```
(you have to copy the snippet verbatim including the whitespace as it is!) you get a much better diff:

```diff
             <module>
-                <name>user_first</name>
+                <label>Installation Mode</label>
+                <name>sap-start</name>
+                <enable_back>yes</enable_back>
+                <enable_next>yes</enable_next>
             </module>
```

## Change Only the Node Text

You can simplify the whole tag matches like

```xml
<xsl:template match="n:try_separate_home">
  <try_separate_home config:type="boolean">false</try_separate_home>
</xsl:template>
```

to just match for the inner text:

```xml
<xsl:template match="n:try_separate_home/text()">false</xsl:template>
```

This avoids repeating whole tags, you do not have to take care of the `config:type` attribute as this will replace only the inner content. So there is less chance for errors...

## Diff Hint

To make debugging easier I'd suggest to display the diff from the original SLES during the RPM build. This way you can easily verify what has been changed without need to download, unpack and compare the files manually.

I have implemented it in the CaaSP package [here](https://github.com/yast/skelcd-control-CAASP/blob/9c3d9bcb2e8c61093e9a47ca5a87901ecfb05b50/package/skelcd-control-CAASP.spec#L131-L133), you just need to add `BuildRequires:  diffutils` to have `diff` available.

lslezak commented:
The complete XSL file:

```xml
<!--
  Definition of the installation.SLES.xml -> installation.SLES4SAP.xml transformation.
  For now it simply copies all XML tags to the target file.
-->

<xsl:stylesheet version="1.0"
  xmlns:n="http://www.suse.com/1.0/yast2ns"
  xmlns="http://www.suse.com/1.0/yast2ns"
  exclude-result-prefixes="n"
  xmlns:config="http://www.suse.com/1.0/configns"
  xmlns:xsl="http://www.w3.org/1999/XSL/Transform">

<xsl:output method="xml" indent="yes"/>

  <xsl:template match="node()|@*">
    <xsl:copy>
      <xsl:apply-templates select="node()|@*"/>
    </xsl:copy>
  </xsl:template>

  <!-- keep empty <![CDATA[]]>, see https://stackoverflow.com/a/1364900 -->
  <xsl:template match="n:optional_default_patterns">
      <xsl:element name="optional_default_patterns">
          <xsl:text disable-output-escaping="yes">&lt;![CDATA[</xsl:text>
          <xsl:value-of select="text()" disable-output-escaping="yes"/>
          <xsl:text disable-output-escaping="yes">]]&gt;</xsl:text>
      </xsl:element>
  </xsl:template>

  <xsl:template match="n:try_separate_home/text()">false</xsl:template>

  <xsl:template match="n:root_base_size/text()">60GB</xsl:template>

  <xsl:template match="n:root_max_size/text()">80GB</xsl:template>

  <xsl:template xml:space="preserve" match="n:proposal_lvm"><proposal_lvm config:type="boolean">true</proposal_lvm>
	 <lvm_vg_strategy>use_needed</lvm_vg_strategy></xsl:template>

  <xsl:template xml:space="preserve" match="n:module[n:name='user_first']"><module>
                <label>Installation Mode</label>
                <name>sap-start</name>
                <enable_back>yes</enable_back>
                <enable_next>yes</enable_next>
            </module></xsl:template>

</xsl:stylesheet>
```

lslezak commented:
The result it this minimal diff:
```diff
--- installation.SLES.xml	2017-10-05 15:33:00.143694895 +0200
+++ installation.SLES4SAP.xml	2017-10-05 17:20:02.876780617 +0200
@@ -35,12 +35,13 @@
     </texts>
 
     <partitioning>
-        <try_separate_home config:type="boolean">true</try_separate_home>
+        <try_separate_home config:type="boolean">false</try_separate_home>
         <limit_try_home>5GB</limit_try_home>
         <root_space_percent>40</root_space_percent>
-        <root_base_size>3GB</root_base_size>
-        <root_max_size>10GB</root_max_size>
-        <proposal_lvm config:type="boolean">false</proposal_lvm>
+        <root_base_size>60GB</root_base_size>
+        <root_max_size>80GB</root_max_size>
+        <proposal_lvm config:type="boolean">true</proposal_lvm>
+	 <lvm_vg_strategy>use_needed</lvm_vg_strategy>
         <vm_desired_size>15GB</vm_desired_size>
         <vm_home_max_size>25GB</vm_home_max_size>
         <btrfs_increase_percentage config:type="integer">300</btrfs_increase_percentage>
@@ -212,7 +213,10 @@
                 </arguments>
             </module>
             <module>
-                <name>user_first</name>
+                <label>Installation Mode</label>
+                <name>sap-start</name>
+                <enable_back>yes</enable_back>
+                <enable_next>yes</enable_next>
             </module>
             <module>
                 <name>root_first</name>

```

lslezak commented:
And to fix the Travis failure you have to additionally install the SLES package, add this line just after the `FROM` line in the `Dockerfile`:

```
RUN zypper --non-interactive in --no-recommends skelcd-control-SLES
```

lslezak commented on package/skelcd-control-SLES4SAP.spec:
```diff
@@ -60,7 +60,8 @@ SLES4SAP control file needed for installation
 # transform ("patch") the original SLES installation file
-xsltproc %{SOURCE0} /installation.xml > installation.xml
+xsltproc %{SOURCE0} /installation.xml > installation1.xml
+xmllint --format installation1.xml > installation.xml
```
With my proposed XSL you do not need to reformat...

lslezak reviewed (changes requested):
See the proposed changes above, also fix the Travis build issue.

varkoly committed:
Replace the whole partitioning section.

lslezak commented:
Oh, I just noticed that the header still contains `For now it simply copies all XML tags to the target file.` which is not true anymore, it **does** some changes...

varkoly committed:
Insert sap_inst in second stage to be able to start yast2-firstboot

varkoly committed:
Apply the suggestions of lslezak

varkoly commented:
Hi Ladislav, thank you for your suggestions. I think I've applied all right.
But I fear that xml:space="preserve" does not really works.
Furthermore  I've replaced the whole <partitioning> because old and new tags must not be mixed.

varkoly commented:
I do not know how to fix travis  build error.
skelcd-control-SLES4SAP.spec contains:
BuildRequires: skelcd-control-SLES >= 15.0.0

And:
rpm -qlp skelcd-control-SLES-15.0.7-1.1.x86_64.rpm
warning: skelcd-control-SLES-15.0.7-1.1.x86_64.rpm: Header V3 RSA/SHA256 Signature, key ID 39db7c82: NOKEY
/installation.xml
/usr/share/doc/packages/skelcd-control-SLES
/usr/share/doc/packages/skelcd-control-SLES/LICENSE

Why does not exist /installation.xml?

When I build it locally it works fine expect of the check of the new installation.xml.
This can not work because the actual /usr/share/YaST2/control/control.rng does not contains the new definition partitioning tags.

varkoly committed:
Fix Travis failure

varkoly committed:
Adapt Ladislav's suggestion

lslezak commented:
Thanks for improving it!

Um, the `xml:space="preserve"` works for me, maybe it depends on the XSL processor version...

Anyway, if you change the partitioning completely then I'd probably put the new values into a separate XML document. So you do not mix the XSL processing elements with the replaced XML data and make the XSL file smaller and better readable.

See the example in the [control.openSUSE-promo.xsl](https://github.com/yast/skelcd-control-openSUSE/blob/master/control/control.openSUSE-promo.xsl#L20) which replaces the `products_supported_for_upgrade` node with the content stored in a separate [control.openSUSE-promo.changes.xml](https://github.com/yast/skelcd-control-openSUSE/blob/master/control/control.openSUSE-promo.changes.xml). I think it would be a good idea to use the same for the partitioning here...

varkoly committed:
Include sles4sap role

varkoly committed:
Merge branch 'master' of github.com:varkoly/skelcd-control-SLES4SAP

varkoly committed:
Fix spec

lslezak commented on package/skelcd-control-SLES4SAP.spec:
```diff
@@ -61,6 +62,9 @@ SLES4SAP control file needed for installation
 xsltproc %{SOURCE0} /installation.xml > installation.xml
+#xsltproc %{SOURCE0} /installation.xml > installation1.xml
+#xmllint --format installation1.xml > installation.xml
+diff -u /installation.xml > installation.xml || :
```
There is extra `>` character.

lslezak reviewed (changes requested):
Wrong `diff` command

varkoly committed:
- bsc#1066678 New partitioning proposal does not work.
  Fixed the fs tags. From <fstypes> to <fs_type>
- Fix swap size
- Remove weight
- FATE#323792: [MODULE] Create SAP Applications Module
  Applying the new partitioning proposal.
- Start inst_sap-start client via <additional_dialogs>
- Adapt pattern names

varkoly committed:
Introduce sap_server pattern again.

varkoly committed:
Turn on xml check.

varkoly committed:
New Version

lslezak reviewed (approved):
LGTM

Commit:
Adapt to Devel:SAP:SLE-15:GA

Commit:
Adapt to Devel:SAP:SLE-15:GA

Commit:
Adapt to Devel.

varkoly committed:
Merge with build system

varkoly committed:
Merge branch 'master' of github.com:varkoly/skelcd-control-SLES4SAP

varkoly commented:
Please merge it. I can not do it.
